src/evaluate.js

```diff
@@ -0,0 +1,379 @@
+/* eslint-disable no-new-func */
+
+'use strict';
+
+var ndloop = require('./utils/ndloop');
+var variable = require('./utils/variable');
+var accessorPreamble = require('./utils/accessor-preamble');
+var inferType = require('./utils/infer-type');
+var isArrayLike = require('./utils/is-array-like');
+
+var evaluatorCache = {};
+var codeCache = {};
+
+module.exports = function (cacheKey, nurbs, accessors, debug, checkBounds, isBasis, derivative, derivativeDimension) {
+  if (isBasis) cacheKey = 'Basis' + cacheKey;
+  if (derivative) cacheKey = 'Der' + derivative + '_' + derivativeDimension + '_' + cacheKey;
+  var cachedEvaluator = evaluatorCache[cacheKey];
+  if (debug) {
+    var logger = typeof debug === 'function' ? debug : console.log;
+  }
+
+  if (cachedEvaluator) {
+    if (debug) {
+      logger(codeCache[cacheKey]);
+    }
+
+    return cachedEvaluator.bind(nurbs);
+  }
+
+  var points = nurbs.points;
+  var degree = nurbs.degree;
+  var weights = nurbs.weights;
+  var knots = nurbs.knots;
+  var splineDimension = nurbs.splineDimension;
+  var spaceDimension = nurbs.dimension;
+  var boundary = nurbs.boundary;
+
+  var i, j, n, m, d, kvar;
+  var code = [];
+  var functionName = 'evaluate' + cacheKey;
+
+  var pointAccessor = accessors.point;
+  if (isBasis) {
+    pointAccessor = function (src, period) {
+      var terms = [];
+      for (var i = 0; i < src.length; i++) {
+        var accessor = src[i];
+        var terms2 = [];
+        for (var j = 0; j < accessor.length; j++) {
+          if (accessor[j] !== 0) terms2.push(accessor[j]);
+        }
+        accessor = terms2.join(' + ');
+        if (period[i]) {
+          accessor = '(' + accessor + ' + ' + period[i] + ') % ' + period[i];
+        }
+        terms.push(accessor + ' === ' + indexVar(i));
+      }
+      return '((' + terms.join(' && ') + ') ? 1 : 0)';
+    };
+  }
+  var weightAccessor = accessors.weight;
+  var knotAccessor = accessors.knot;
+
+  var knotVar = variable('k');
+  var pointVar = variable('x');
+  var weightVar = variable('w');
+  var indexVar = variable('i');
+  var tVar = variable('t');
+  var domainVar = debug ? 'domain' : 'd';
+  var sizeVar = variable(debug ? 'size' : 's');
+  var knotIndex = variable(debug ? 'knotIndex' : 'j');
+
+  var hasWeights = weights !== undefined;
+  var allDimensionUniform = true;
+  for (d = 0; d < splineDimension; d++) {
+    if (isArrayLike(knots) && isArrayLike(knots[d])) {
+      allDimensionUniform = false;
+    }
+  }
+
+  // Just to indent properly and save lots of typing
+  function line (str) {
+    code.push('  ' + (str || ''));
+  }
+  function debugLine (str) {
+    if (debug) line(str);
+  }
+  // function clog (str) {
+    // if (debug) code.push('console.log("' + str + ' =", ' + str + ');');
+  // }
```

Comment:
These lines don't seem to be in use

Reply:
Correct. I kept rewriting this function because I kept deleting it to clean things up. So I finally just left it in. It's really very useful for development/debugging.